tensorboard/components_polymer3/tf_line_chart_data_loader/tf-line-chart-data-loader.ts

```diff
@@ -0,0 +1,289 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {html} from '@polymer/polymer';
+import * as Plottable from 'plottable';
+import {customElement, property, observe} from '@polymer/decorators';
+import * as _ from 'lodash';
+import '@polymer/paper-spinner/paper-spinner';
+
+import {runsColorScale} from '../tf_color_scale/colorScale';
+import {RequestManager} from '../tf_backend/requestManager';
+import {
+  XType,
+  TooltipColumn,
+  SymbolFn,
+  ScalarDatum,
+} from '../vz_chart_helpers/vz-chart-helpers';
+import {DataLoader} from '../tf_dashboard_common/data-loader-behavior';
+import {VzLineChart2} from '../vz_line_chart2/vz-line-chart2';
+import {YScaleType, FillArea} from '../vz_line_chart2/line-chart';
+
+// The chart can sometimes get in a bad state, when it redraws while
+// it is display: none due to the user having switched to a different
+// page. This code implements a cascading queue to redraw the bad charts
+// one-by-one once they are active again.
+// We use a cascading queue becuase we don't want to block the UI / make the
+// ripples very slow while everything synchronously redraws.
+const redrawQueue: TfLineChartDataLoader<any>[] = [];
+let redrawRaf = 0;
+const cascadingRedraw = _.throttle(function internalRedraw() {
+  if (redrawQueue.length == 0) return;
+  const x = redrawQueue.shift();
+  if (x && x.active) {
+    x.redraw();
+    x._maybeRenderedInBadState = false;
+  }
+  window.cancelAnimationFrame(redrawRaf);
+  window.requestAnimationFrame(internalRedraw);
+}, 100);
+
+// A component that fetches data from the TensorBoard server and renders it into
+// a vz-line-chart.
+@customElement('tf-line-chart-data-loader')
+export class TfLineChartDataLoader<ScalarMetadata> extends DataLoader<
```

Comment:
Question - where is the `ScalarMetadata` type coming from?  I'm getting an import error when I'm trying to parameterize a TfLineChartDataLoader<> for the scalar dashboard (with your PR patched locally).

Reply:
Or I see, that's actually a type parameter?  So I guess I need to define the appropriate type on my end.

Reply:
Yup, it is a generic.